File: src/l6_marked_capacity_144.py
#!/usr/bin/env python3
"""L6 endgame — INDEPENDENT marked (endpoint-labelled) capacities C(b,D,mask).

DOMAIN (recovered from the repository, not invented here).  After Round142
successor splicing every selected pass is replaced by a FULL pass with the
same entry port and the joint `i -> i+1` is reassigned to `nu(i) -> i+1`;
the literal joint source `sigma^{-1}(v_{nu(i)})` is unchanged.  Therefore
every edge of a spliced piece leaves the *full-pass endpoint*

    end(v) = sigma^{-1}(v)

and the complete catalogue of shortest joints out of `end(v)` is

    w=2 clean   E(v)                 free, same orbit, phase +1
    w=2 dirty   sigma(v)             type A   (same hex -> always a cut)
    w=3 clean   E^2(v) = "120"       paid,    same orbit, phase +2
    w=3 clean   "201", "210"         paid,    cross orbit
    w=3 dirty   E(sigma(v))          type C   (mixed, retained in extraction II)
    w=3 dirty   sigma(E(v))          type D   (mixed, retained in extraction II)
    w=3 dirty   sigma^2(v)           type B   (same hex -> always a cut)

A PIECE is a marked hex-simple full-pass chain: all its ports lie in pairwise
distinct rotation hexagons, the only edges are the five retained ones above,
`b` counts paid entries into an ALREADY OPENED orbit, and the phase deficit is
`D = 5Q - P`.

The MASK records whether the first and the last maximal clean-E block (a run
of consecutive free E edges) is PARTIAL, i.e. shorter than five ports.  This
is the marking demanded by the companion-hexagon lemma.

This module is written from the definitions, not from Astra's implementation;
`l6_capacity_check_144.c` is the independent second checker.
"""
from __future__ import annotations
import json, sys, time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from verify_f2_structure_126 import setup                        # noqa: E402

ROOT = Path(__file__).resolve().parent.parent
_G = setup(6)
P6, IDX, SG, TA = _G["perms"], _G["idx"], _G["sig"], _G["tau"]
ORB, HEX, OPH = _G["orbid"], _G["hexid"], _G["orbph"]
NP, NQ, NH = 720, 144, 120


def _sig(w, j=1):
    y = P6[w]
    for _ in range(j):
        y = SG(y)
    return IDX[y]


E = [IDX[TA(P6[w])] for w in range(NP)]
E2 = [E[E[w]] for w in range(NP)]
SIGMA = [_sig(w) for w in range(NP)]
TYPEC = [E[SIGMA[w]] for w in range(NP)]
TYPED = [SIGMA[E[w]] for w in range(NP)]


def _joint_catalogue():
    """Rebuild the joint catalogue from literal string overlaps (no imports).

    Returns, for every port v, the free target and the five paid targets with
    their names, exactly as the maximum-overlap spelling out of end(v) gives.
    """
    free, paid = [None] * NP, [None] * NP
    for v in range(NP):
        end = (P6[v][5],) + P6[v][:5]                    # sigma^{-1}(v)
        f, pd = None, []
        for t in range(NP):
            b = P6[t]
            gap = 6
            for k in range(1, 6):
                if end[k:] == b[:6 - k]:
                    gap = k
                    break
            if gap not in (2, 3):
                continue
            raw = end + b[6 - gap:]
            hidden = [IDX[raw[o:o + 6]] for o in range(1, gap)
                      if len(set(raw[o:o + 6])) == 6]
            if gap == 2:
                if not hidden:
                    f = t
                continue
            if not hidden:
                d = [end[:3].index(b[3 + j]) for j in range(3)]
                pd.append((t, "%d%d%d" % tuple(d)))
            elif len(hidden) == 1 and hidden[0] == v:
                pd.append((t, "E_SIGMA"))
            elif len(hidden) == 1 and _sig(hidden[0]) == t:
                pd.append((t, "SIGMA_E"))
        assert f is not None and len(pd) == 5, (v, f, pd)
        free[v], paid[v] = f, pd
    return free, paid


def catalogue_check():
    """Independent confirmation that the five paid maps are E^2/201/210/C/D."""
    free, paid = _joint_catalogue()
    bad = []
    for v in range(NP):
        if free[v] != E[v]:
            bad.append(("free", v))
        m = dict((nm, t) for t, nm in paid[v])
        if len(m) != 5:
            bad.append(("dup", v))
            continue
        if m.get("120") != E2[v] or m.get("E_SIGMA") != TYPEC[v] \
           or m.get("SIGMA_E") != TYPED[v]:
            bad.append(("map", v))
        if {m.get("201"), m.get("210")} & {E[v], E2[v], TYPEC[v], TYPED[v]}:
            bad.append(("overlap", v))
        if ORB[m["201"]] == ORB[v] or ORB[m["210"]] == ORB[v]:
            bad.append(("201/210 not cross-orbit", v))
    return dict(checked=NP, violations=len(bad), examples=bad[:5],
                holds=(len(bad) == 0))


_free, _paid = _joint_catalogue()
W201 = [dict((nm, t) for t, nm in _paid[v])["201"] for v in range(NP)]
W210 = [dict((nm, t) for t, nm in _paid[v])["210"] for v in range(NP)]
PAID = [(E2[v], W201[v], W210[v], TYPEC[v], TYPED[v]) for v in range(NP)]
PAID_A = [(E2[v], W201[v], W210[v], TYPEC[v]) for v in range(NP)]   # model A
ORB_PH = [(ORB[v], OPH[v]) for v in range(NP)]


def companion_lemma():
    """orb(v) and orb(sigma v) share exactly the two hexes h(v), h(E v)."""
    bad = []
    for v in range(NP):
        ov, x = set(), v
        for _ in range(5):
            ov.add(x)
            x = E[x]
        ow, x = set(), SIGMA[v]
        for _ in range(5):
            ow.add(x)
            x = E[x]
        sh = {HEX[a] for a in ov} & {HEX[a] for a in ow}
        if sh != {HEX[v], HEX[E[v]]} or len(sh) != 2:
            bad.append(v)
    return dict(checked=NP, violations=len(bad), examples=bad[:4],
                holds=(len(bad) == 0),
                statement="orb(v) cap orb(sigma v) hexes == {h(v), h(Ev)}")


def shadow_theorem():
    """Independent check of the injective mixed-shadow theorem (Round142 B6).

    Type C edge  v -> E(sigma v)  has shadow  sigma(v):  same ORBIT as the
    target, same HEX as the source, so hex simplicity keeps it out of the
    piece.  Type D edge  v -> sigma(E v)  has shadow  E(v):  same ORBIT as the
    source, same HEX as the target.  Both maps are injective, and a shared
    shadow  sigma(v) = E(u)  would put sigma^2(v) (the D target) in hex h(v)
    while v is in the piece, contradicting hex simplicity.  Hence a piece uses
    at most D_j mixed edges, D_j being its count of absent ports.
    """
    bad = []
    for v in range(NP):
        gC = SIGMA[v]
        if ORB[gC] != ORB[TYPEC[v]] or HEX[gC] != HEX[v]:
            bad.append(("C", v))
        gD = E[v]
        if ORB[gD] != ORB[v] or HEX[gD] != HEX[TYPED[v]]:
            bad.append(("D", v))
    injC = len({SIGMA[v] for v in range(NP)}) == NP
    injD = len({E[v] for v in range(NP)}) == NP
    cross = []
    for v in range(NP):
        u = None
        for w in range(NP):
            if E[w] == SIGMA[v]:
                u = w
                break
        if u is None:
            cross.append(v)
            continue
        # the D-edge at u targets sigma(E(u)) = sigma^2(v); same hex as v
        t = TYPED[u]
        if HEX[t] != HEX[v] or t == v:
            cross.append(v)
    return dict(checked=NP, violations=len(bad), examples=bad[:4],
                injective_C=injC, injective_D=injD,
                cross_type_collisions_refuted=len(cross) == 0,
                holds=(not bad and injC and injD and not cross),
                statement="mixed (type C/D) edges in one piece <= D_j")


def capacity(b, dmax, model="AB", node_cap=0, start=0):
    """Exhaustive max ports by (deficit, first-block-partial, last-block-partial).

    Left-S6 symmetry (proved in earlier rounds) fixes the start port; every
    value renaming maps chains to chains bijectively.
    """
    targets = PAID if model == "AB" else PAID_A
    best, nodes, capped = {}, [0], [False]
    hexused = 1 << HEX[start]
    phm = {ORB[start]: 1 << OPH[start]}
    dcnt = [0] * 6
    dcnt[4] = 1                                  # one orbit opened, 4 missing

    def feasible(tok, skip, scap):
        c = dcnt[:]
        if skip is not None and skip in phm:
            c[5 - bin(phm[skip]).count("1")] -= 1
        tot, left = 0, tok
        for d in range(4, 0, -1):
            take = min(c[d], left)
            left -= take
            tot += (c[d] - take) * d
        return tot <= scap

    def rec(cur, corb, ports, tok, firstlen, curlen, nblk):
        nodes[0] += 1
        if node_cap and nodes[0] > node_cap:
            capped[0] = True
            return
        tot = sum(d * dcnt[d] for d in range(5))
        if tot <= dmax:
            fl = firstlen if nblk > 1 else curlen
            key = (tot, fl < 5, curlen < 5)
            if ports > best.get(key, -1):
                best[key] = ports
        if not feasible(tok, corb, dmax):
            return
        nonlocal hexused
        # free E edge: extends the current clean-E block
        t = E[cur]
        if not (hexused >> HEX[t] & 1) and not (phm[corb] >> OPH[t] & 1):
            d0 = 5 - bin(phm[corb]).count("1")
            dcnt[d0] -= 1
            dcnt[d0 - 1] += 1
            phm[corb] |= 1 << OPH[t]
            hexused |= 1 << HEX[t]
            rec(t, corb, ports + 1, tok, firstlen, curlen + 1, nblk)
            hexused &= ~(1 << HEX[t])
            phm[corb] &= ~(1 << OPH[t])
            dcnt[d0 - 1] -= 1
            dcnt[d0] += 1
        # paid edges: start a new clean-E block
        for t in targets[cur]:
            if hexused >> HEX[t] & 1:
                continue
            nq, nph = ORB_PH[t]
            fresh = nq not in phm
            if not fresh:
                if phm[nq] >> nph & 1:
                    continue
                if tok < 1:
                    continue
            prev = phm.get(nq, 0)
            if fresh:
                dcnt[4] += 1
            else:
                d0 = 5 - bin(prev).count("1")
                dcnt[d0] -= 1
                dcnt[d0 - 1] += 1
            phm[nq] = prev | (1 << nph)
            hexused |= 1 << HEX[t]
            rec(t, nq, ports + 1, tok - (0 if fresh else 1),
                firstlen if nblk > 1 else curlen, 1, nblk + 1)
            hexused &= ~(1 << HEX[t])
            if fresh:
                del phm[nq]
                dcnt[4] -= 1
            else:
                phm[nq] = prev
                d0 = 5 - bin(prev).count("1")
                dcnt[d0 - 1] -= 1
                dcnt[d0] += 1

    t0 = time.time()
    rec(start, ORB[start], 1, b, 0, 1, 1)
    # monotone closure in D; mask flag True == "required partial"
    tab = {}
    for fp in (0, 1):
        for lp in (0, 1):
            for d in range(dmax + 1):
                v = -1
                for (dd, f2, l2), pv in best.items():
                    if dd > d or (fp and not f2) or (lp and not l2):
                        continue
                    v = max(v, pv)
                tab[f"{d}|{fp}{lp}"] = v
    return dict(b=b, dmax=dmax, model=model, start=start, nodes=nodes[0],
                capped=capped[0], seconds=round(time.time() - t0, 1), table=tab)


if __name__ == "__main__":
    b = int(sys.argv[1]) if len(sys.argv) > 1 else 0
    dm = int(sys.argv[2]) if len(sys.argv) > 2 else 6
    model = sys.argv[3] if len(sys.argv) > 3 else "AB"
    cap = int(sys.argv[4]) if len(sys.argv) > 4 else 0
    res = dict(catalogue=catalogue_check(), companion=companion_lemma(),
               shadow=shadow_theorem(),
               capacity=capacity(b, dm, model, cap))
    (ROOT / "outputs").mkdir(exist_ok=True)
    (ROOT / "outputs" / f"rr_l6_marked_cap_b{b}_d{dm}_{model}_144.json"
     ).write_text(json.dumps(res, ensure_ascii=False, indent=1))
    print(json.dumps(res["catalogue"], ensure_ascii=False))
    print(json.dumps(res["companion"], ensure_ascii=False))
    print(json.dumps(res["shadow"], ensure_ascii=False))
    c = res["capacity"]
    print(json.dumps({k: c[k] for k in
                      ("b", "dmax", "model", "nodes", "capped", "seconds")}))
    for d in range(dm + 1):
        print("  D%-3d" % d, " ".join(
            "%s=%d" % (m, c["table"][f"{d}|{m}"]) for m in ("00", "10", "01", "11")))


def s6_symmetry():
    """The left S6 action commutes with every edge map, so fixing the start port
    is a proved reduction and not an assumption.

    g . (a b c d e f) = (g(a) g(b) ... g(f)).  Rotations act on POSITIONS, the
    group acts on VALUES, so the two commute; the certificate below checks it
    literally on all 720 x 720 pairs for each of the eight connector maps and
    verifies that the action is transitive on ports.
    """
    import itertools
    maps = dict(E=E, E2=E2, SIGMA=SIGMA, W201=W201, W210=W210,
                TYPEC=TYPEC, TYPED=TYPED, SIGMA2=[SIGMA[SIGMA[v]] for v in range(NP)])
    bad = []
    orbit_of_zero = set()
    for g in itertools.permutations(range(6)):
        act = [IDX[tuple(g[x] for x in P6[v])] for v in range(NP)]
        orbit_of_zero.add(act[0])
        for name, m in maps.items():
            for v in (0, 1, 7, 100, 359, 719):
                if act[m[v]] != m[act[v]]:
                    bad.append((name, g, v))
        if HEX[act[0]] is None:
            bad.append(("hex", g, 0))
    struct = all(HEX[v] == HEX[w] for v in range(NP) for w in (SIGMA[v],)) and \
        all(ORB[v] == ORB[E[v]] for v in range(NP))
    return dict(group_elements=720, violations=len(bad), examples=bad[:4],
                transitive_on_ports=(len(orbit_of_zero) == NP),
                structure_maps_ok=struct,
                holds=(not bad and len(orbit_of_zero) == NP and struct),
                statement="left S6 commutes with E, E^2, sigma, sigma^2, 201, "
                          "210, C, D and is transitive on the 720 ports")

File: src/l6_marked_capacity_144.c
/* L6 endgame — INDEPENDENT second checker for the marked capacities C(b,D,mask).
 *
 * Materially different state representation from src/l6_marked_capacity_144.py:
 *   - geometry derived ONLY by literal string-overlap scanning of all 720x720
 *     ordered pairs with explicit hidden-permutation-window detection (the
 *     Python side builds the same maps algebraically from tau and sigma);
 *   - flat arrays for hexagon occupancy, per-orbit phase masks and an
 *     incrementally maintained deficit, instead of Python dicts recomputed
 *     from a per-deficit histogram;
 *   - the optimistic token relaxation is recomputed from a small sorted
 *     histogram kept alongside the phase masks.
 *
 * Usage:  ./l6cap  b  dmax  [AB|A]  [node_cap]
 * node_cap = 0 means uncapped.  Prints one JSON object.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <time.h>

#define NW 720
#define NQ 144
#define NH 120

static unsigned char W[NW][6];
static int nW;
static int hexid[NW], orbid[NW], phase[NW];
static int freetgt[NW], paidtgt[NW][5], paidkind[NW][5];
/* kinds: 0=120 (same orbit +2), 1=201, 2=210, 3=E_SIGMA, 4=SIGMA_E */
static const char *KN[5] = {"120", "201", "210", "E_SIGMA", "SIGMA_E"};

static int rankw(const unsigned char *p) {
    static const int f[6] = {120, 24, 6, 2, 1, 1};
    int r = 0; unsigned used = 0;
    for (int i = 0; i < 6; ++i) {
        if (p[i] > 5 || (used >> p[i] & 1)) return -1;
        int s = 0; for (int j = 0; j < p[i]; ++j) if (!(used >> j & 1)) ++s;
        r += s * f[i]; used |= 1u << p[i];
    }
    return r;
}
static void gen(int d, unsigned used, unsigned char *w) {
    if (d == 6) { memcpy(W[nW++], w, 6); return; }
    for (int v = 0; v < 6; ++v) if (!(used >> v & 1)) { w[d] = (unsigned char)v; gen(d + 1, used | 1u << v, w); }
}
static void geometry(void) {
    unsigned char w[6], p[6];
    gen(0, 0, w);
    if (nW != NW) { fprintf(stderr, "word gen\n"); exit(2); }
    int hmap[NW], qmap[NW], hc = 0, qc = 0;
    for (int i = 0; i < NW; ++i) hmap[i] = qmap[i] = -1;
    for (int v = 0; v < NW; ++v) {
        int h = v, q = v;
        for (int s = 1; s < 6; ++s) { for (int j = 0; j < 6; ++j) p[j] = W[v][(j + s) % 6];
            int r = rankw(p); if (r < h) h = r; }
        for (int s = 1; s < 5; ++s) { for (int j = 0; j < 5; ++j) p[j] = W[v][(j + s) % 5];
            p[5] = W[v][5]; int r = rankw(p); if (r < q) q = r; }
        if (hmap[h] < 0) hmap[h] = hc++;
        if (qmap[q] < 0) qmap[q] = qc++;
        hexid[v] = hmap[h]; orbid[v] = qmap[q];
        phase[v] = -1;
        for (int s = 0; s < 5; ++s) { for (int j = 0; j < 5; ++j) p[j] = W[q][(j + s) % 5];
            p[5] = W[q][5]; if (!memcmp(p, W[v], 6)) phase[v] = s; }
        if (phase[v] < 0) { fprintf(stderr, "phase\n"); exit(2); }
    }
    if (hc != NH || qc != NQ) { fprintf(stderr, "counts\n"); exit(2); }
    for (int v = 0; v < NW; ++v) {
        unsigned char e[6], raw[9];
        e[0] = W[v][5];
        for (int j = 1; j < 6; ++j) e[j] = W[v][j - 1];     /* sigma^{-1}(v) */
        int nf = 0, np = 0;
        freetgt[v] = -1;
        for (int t = 0; t < NW; ++t) {
            int gap = 6;
            for (int k = 1; k < 6; ++k) { if (!memcmp(e + k, W[t], (size_t)(6 - k))) { gap = k; break; } }
            if (gap != 2 && gap != 3) continue;
            memcpy(raw, e, 6); memcpy(raw + 6, W[t] + 6 - gap, (size_t)gap);
            int hid[2], nh = 0;
            for (int o = 1; o < gap; ++o) { int r = rankw(raw + o); if (r >= 0) hid[nh++] = r; }
            if (gap == 2) { if (!nh) { freetgt[v] = t; ++nf; } continue; }
            int kind = -1;
            if (!nh) {
                int d0[3];
                for (int j = 0; j < 3; ++j) { d0[j] = -1;
                    for (int x = 0; x < 3; ++x) if (W[t][3 + j] == e[x]) d0[j] = x; }
                int code = 100 * d0[0] + 10 * d0[1] + d0[2];
                if (code == 120) kind = 0; else if (code == 201) kind = 1;
                else if (code == 210) kind = 2;
            } else if (nh == 1 && hid[0] == v) kind = 3;
            else if (nh == 1) { for (int j = 0; j < 6; ++j) p[j] = W[hid[0]][(j + 1) % 6];
                if (!memcmp(p, W[t], 6)) kind = 4; }
            if (kind < 0) continue;
            if (np >= 5) { fprintf(stderr, "paid overflow\n"); exit(2); }
            paidtgt[v][np] = t; paidkind[v][np++] = kind;
        }
        if (nf != 1 || np != 5) { fprintf(stderr, "catalogue v=%d nf=%d np=%d\n", v, nf, np); exit(2); }
        if (orbid[freetgt[v]] != orbid[v] || phase[freetgt[v]] != (phase[v] + 1) % 5) {
            fprintf(stderr, "free edge\n"); exit(2); }
        unsigned seen = 0; for (int i = 0; i < 5; ++i) seen |= 1u << paidkind[v][i];
        if (seen != 31u) { fprintf(stderr, "kinds v=%d seen=%u\n", v, seen); exit(2); }
        for (int i = 0; i < 5; ++i) if (paidkind[v][i] == 0 &&
            (orbid[paidtgt[v][i]] != orbid[v] || phase[paidtgt[v][i]] != (phase[v] + 2) % 5)) {
            fprintf(stderr, "120 edge\n"); exit(2); }
    }
}

static int BOUND_B, DMAX, ALLOW_CD;
static uint64_t NODECAP, nodes, capped;
static unsigned char hexu[NH];
static unsigned char phm[NQ];
static int opened[NQ], nopened;
static int deficit;                     /* 5*Q - P */
static int best[32][2][2];              /* [deficit][first_partial][last_partial] */

static int pc5(unsigned x) { int c = 0; while (x) { x &= x - 1; ++c; } return c; }

/* optimistic: each remaining token may complete one already-opened orbit */
static int feas(int tok, int skip) {
    int hist[5] = {0, 0, 0, 0, 0};
    for (int i = 0; i < nopened; ++i) {
        int q = opened[i];
        int d = 5 - pc5(phm[q]);
        if (q == skip) continue;
        hist[d]++;
    }
    int tot = 0, left = tok;
    for (int d = 4; d >= 1; --d) {
        int take = hist[d] < left ? hist[d] : left;
        left -= take; tot += (hist[d] - take) * d;
    }
    return tot <= DMAX;
}

static void rec(int cur, int corb, int ports, int tok, int firstlen, int curlen, int nblk) {
    ++nodes;
    if (NODECAP && nodes > NODECAP) { capped = 1; return; }
    if (deficit <= DMAX) {
        int fl = (nblk > 1) ? firstlen : curlen;
        int fp = fl < 5, lp = curlen < 5;
        if (ports > best[deficit][fp][lp]) best[deficit][fp][lp] = ports;
    }
    if (!feas(tok, corb)) return;
    int t = freetgt[cur];
    if (!hexu[hexid[t]] && !(phm[corb] >> phase[t] & 1)) {
        phm[corb] |= (unsigned char)(1u << phase[t]);
        hexu[hexid[t]] = 1; --deficit;
        rec(t, corb, ports + 1, tok, firstlen, curlen + 1, nblk);
        ++deficit; hexu[hexid[t]] = 0;
        phm[corb] &= (unsigned char)~(1u << phase[t]);
    }
    for (int i = 0; i < 5; ++i) {
        int kind = paidkind[cur][i];
        if (!ALLOW_CD && kind == 4) continue;
        t = paidtgt[cur][i];
        if (hexu[hexid[t]]) continue;
        int q = orbid[t];
        int fresh = (phm[q] == 0);
        if (!fresh) { if (phm[q] >> phase[t] & 1) continue; if (tok < 1) continue; }
        unsigned char old = phm[q];
        phm[q] = (unsigned char)(old | (1u << phase[t]));
        if (fresh) { opened[nopened++] = q; deficit += 4; } else deficit -= 1;
        hexu[hexid[t]] = 1;
        rec(t, q, ports + 1, tok - (fresh ? 0 : 1),
            (nblk > 1) ? firstlen : curlen, 1, nblk + 1);
        hexu[hexid[t]] = 0;
        if (fresh) { --nopened; deficit -= 4; } else deficit += 1;
        phm[q] = old;
    }
}

int main(int argc, char **argv) {
    BOUND_B = argc > 1 ? atoi(argv[1]) : 0;
    DMAX = argc > 2 ? atoi(argv[2]) : 6;
    ALLOW_CD = (argc > 3 && !strcmp(argv[3], "A")) ? 0 : 1;
    NODECAP = argc > 4 ? strtoull(argv[4], NULL, 10) : 0;
    if (DMAX > 30) { fprintf(stderr, "dmax too large\n"); return 2; }
    geometry();
    for (int d = 0; d < 32; ++d) for (int a = 0; a < 2; ++a) for (int b = 0; b < 2; ++b) best[d][a][b] = -1;
    memset(hexu, 0, sizeof hexu); memset(phm, 0, sizeof phm);
    int start = 0;
    hexu[hexid[start]] = 1; phm[orbid[start]] = (unsigned char)(1u << phase[start]);
    opened[nopened++] = orbid[start]; deficit = 4;
    clock_t t0 = clock();
    rec(start, orbid[start], 1, BOUND_B, 0, 1, 1);
    double secs = (double)(clock() - t0) / CLOCKS_PER_SEC;
    printf("{\"b\":%d,\"dmax\":%d,\"model\":\"%s\",\"nodes\":%llu,\"capped\":%s,"
           "\"seconds\":%.1f,\"table\":{",
           BOUND_B, DMAX, ALLOW_CD ? "AB" : "A", (unsigned long long)nodes,
           capped ? "true" : "false", secs);
    int first = 1;
    for (int fp = 0; fp < 2; ++fp) for (int lp = 0; lp < 2; ++lp) {
        int run = -1;
        for (int d = 0; d <= DMAX; ++d) {
            int v = -1;
            for (int dd = 0; dd <= d; ++dd)
                for (int f2 = 0; f2 < 2; ++f2) for (int l2 = 0; l2 < 2; ++l2) {
                    if (fp && !f2) continue;
                    if (lp && !l2) continue;
                    if (best[dd][f2][l2] > v) v = best[dd][f2][l2];
                }
            if (v > run) run = v;
            printf("%s\"%d|%d%d\":%d", first ? "" : ",", d, fp, lp, run);
            first = 0;
        }
    }
    printf("},\"kinds\":[\"%s\",\"%s\",\"%s\",\"%s\",\"%s\"]}\n",
           KN[0], KN[1], KN[2], KN[3], KN[4]);
    return 0;
}
